### PR TITLE
@ #493 | should validate plugin params before proceeding

### PR DESCRIPTION
**Pull request by hieptranquoc**
connect #493 
please review this PR!

### CODE REVIEW

**Comment by hieptranquoc** [comment]
@hoatle I checked every case, and now the plugin configuration will work properly.

File: lib/teracy-dev/plugin.rb [hunk 1]
```diff
@@ -15,24 +15,38 @@ def self.sync(plugins)
...
-              "https://rubygems.org/",
-              "https://gems.hashicorp.com/"
-            ]
-          end
-
-          plugin_manager.install_plugin(plugin['name'], Util.symbolize(plugin))
-          reload_required = true
+        name_validate = validate(plugin, 'name' => plugin['name'])
+        if !name_validate
+          logger.warn("Plugin name must be configured, not nil, and not empty")
```
Comment (datphan) [inline lib/teracy-dev/plugin.rb]: "Plugin name must be configured"

File: lib/teracy-dev/plugin.rb [hunk 2]
```diff
@@ -15,24 +15,38 @@ def self.sync(plugins)
...
-        if installed_plugins.has_key?(plugin['name']) and plugin['state'] == 'uninstalled'
-          logger.info("uninstalling plugin: #{plugin['name']}")
-          plugin_manager.uninstall_plugin(plugin['name'])
-          reload_required = true
+        case plugin['state']
+        when 'installed'
+          if installed_plugins.empty?
+            source_validate = validate(plugin, 'sources' => plugin['sources'])
+            if !source_validate
+              logger.warn("Plugin sources should be configured when no plugin installed, or we will use default")
```
Comment (datphan) [inline lib/teracy-dev/plugin.rb]: remove this warn

File: lib/teracy-dev/plugin.rb [hunk 3]
```diff
@@ -15,24 +15,38 @@ def self.sync(plugins)
...
+            reload_required = true
+          end
+        when 'uninstalled'
+          if installed_plugins.has_key?(plugin['name']) and plugin['state'] == 'uninstalled'
+            logger.info("uninstalling plugin: #{plugin['name']}")
+            plugin_manager.uninstall_plugin(plugin['name'])
+            reload_required = true
+          end
+        else
+          logger.debug('Plugin state not be set, do nothing')
```
Comment (datphan) [inline lib/teracy-dev/plugin.rb]: when plugin.enabled is `true` and `state` is not setted, we should fill it with `installed` state

**Review by datphan — CHANGES_REQUESTED** [review]

File: lib/teracy-dev/plugin.rb (hunk 3 above)
Comment (hoatle) [reply lib/teracy-dev/plugin.rb]: plugin state is not set, then do nothing is expected. It means that if the plugin is installed or not installed, users don't care.

**Comment by hieptranquoc** [comment]
@datphan I've updated! please check again.

File: lib/teracy-dev/plugin.rb (hunk 3 above)
Comment (datphan) [reply lib/teracy-dev/plugin.rb]: okey, by the way, the message should be "The plugin state is not set, no action will be taken."

**Review by datphan — DISMISSED** [review]

**Review by datphan — DISMISSED** [review]

File: lib/teracy-dev/plugin.rb [hunk 4]
```diff
@@ -45,5 +58,33 @@ def self.sync(plugins)
     def self.installed?(plugin_name)
       return Vagrant.has_plugin?(plugin_name)
     end
+
+    def self.validate(plugin, config)
```
Comment (hoatle) [inline lib/teracy-dev/plugin.rb]: don't introduce this type of complexity, as I can see it just checks the existence. So use Util.exist? instead

File: lib/teracy-dev/plugin.rb [hunk 5]
```diff
@@ -15,24 +15,37 @@ def self.sync(plugins)
...
-            plugin['sources'] = [
-              "https://rubygems.org/",
-              "https://gems.hashicorp.com/"
-            ]
-          end
-
-          plugin_manager.install_plugin(plugin['name'], Util.symbolize(plugin))
-          reload_required = true
+        name_validate = validate(plugin, 'name' => plugin['name'])
+        if !name_validate
```
Comment (hoatle) [inline lib/teracy-dev/plugin.rb]: `unless Util.exist? plugin['name']` is enough for example

**Comment by hieptranquoc** [comment]
@hoatle I've updated! please review again!

**Review by hoatle — APPROVED** [review]
LGTM

**Pull request merged**